File: dragtoclosedialog/src/main/java/com/wengelef/dragtoclosedialog/DragDialogFragment.java
/*
 * Copyright (c) wengelef 2017
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.wengelef.dragtoclosedialog;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.view.MotionEventCompat;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

public abstract class DragDialogFragment extends BaseDialogFragment {

    private static final int DRAG_TO_DISMISS_DISTANCE = 400;

    private float mInitialDragContainerPosition;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View root = inflater.inflate(getLayoutResId(), container, false);

        root.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                close();
            }
        });
        return root;
    }

    @Override
    public void onStart() {
        mAnimationListener = getDialogAnimationListener();
        getContentView().setOnTouchListener(getTouchListener());
        super.onStart();
    }

    private DialogAnimationListener getDialogAnimationListener() {
        return new DialogAnimationListener() {
            @Override
            public void onAnimationEnd() {
                mInitialDragContainerPosition = getContentView().getY();
            }
        };
    }

    private View.OnTouchListener getTouchListener() {
        return new View.OnTouchListener() {

            private float mLastTouchY;
            private float mDistanceCovered;

            @Override
            public boolean onTouch(View v, MotionEvent event) {
                final int action = MotionEventCompat.getActionMasked(event);

                switch (action) {
                    case MotionEvent.ACTION_DOWN: {
                        mLastTouchY = event.getRawY();
                        mDistanceCovered = 0.0f;
                        break;
                    }

                    case MotionEvent.ACTION_MOVE: {
                        final float y = event.getRawY();
                        final float dy = y - mLastTouchY;
                        mDistanceCovered += dy;
                        getContentView().setY(getContentView().getY() + dy);
                        mLastTouchY = y;
                        break;
                    }

                    case MotionEvent.ACTION_UP: {
                        onDragEnd();
                        break;
                    }

                    case MotionEvent.ACTION_CANCEL: {
                        onDragEnd();
                        break;
                    }

                    default: {
                        return true;
                    }
                }
                return true;
            }

            private void onDragEnd() {
                if (mDistanceCovered > DRAG_TO_DISMISS_DISTANCE) {
                    close();
                } else {
                    getContentView().animate().y(mInitialDragContainerPosition);
                }
            }
        };
    }
}

File: dragtoclosedialog/src/main/java/com/wengelef/dragtoclosedialog/DragDialogFragment.kt
package com.wengelef.dragtoclosedialog

import android.support.v4.view.MotionEventCompat
import android.view.MotionEvent
import android.view.View

abstract class DragDialogFragment : BaseDialogFragment() {

    enum class CloseDirection { DEFAULT, BOTTOM, TOP }

    private val DRAG_TO_DISMISS_DISTANCE = 400

    private var initialDragContainerPosition = 0f

    override fun onStart() {
        mAnimationListener = getDialogAnimationListener()
        getContentView().setOnTouchListener(getTouchListener())
        super.onStart()
    }

    protected open fun getCloseDirection(): CloseDirection = CloseDirection.DEFAULT

    private fun getDialogAnimationListener(): DialogAnimationListener = object : DialogAnimationListener {
        override fun onAnimationEnd() {
            initialDragContainerPosition = getContentView().y
        }
    }

    private fun getTouchListener(): View.OnTouchListener = object : View.OnTouchListener {
        private var lastTouchY: Float = 0f
        private var distanceCovered: Float = 0f

        override fun onTouch(v: View, event: MotionEvent): Boolean {
            val action = MotionEventCompat.getActionMasked(event)

            when (action) {
                MotionEvent.ACTION_DOWN -> {
                    lastTouchY = event.rawY
                    distanceCovered = 0f
                }

                MotionEvent.ACTION_MOVE -> {
                    val y = event.rawY
                    val dy = y - lastTouchY
                    distanceCovered += dy
                    getContentView().y = getContentView().y + dy
                    lastTouchY = y
                }

                MotionEvent.ACTION_UP -> onDragEnd()

                MotionEvent.ACTION_CANCEL -> onDragEnd()

                else -> return true
            }
            return true
        }

        private fun onDragEnd() {
            if (getCloseableBottom()) {
                if (distanceCovered > DRAG_TO_DISMISS_DISTANCE) {
                    close()
                } else {
                    getContentView().animate().y(initialDragContainerPosition)
                }
            }
            if (getCloseableTop()) {
                if (distanceCovered < (DRAG_TO_DISMISS_DISTANCE * -1)) {
                    close()
                } else {
                    getContentView().animate().y(initialDragContainerPosition)
                }
            }
        }
    }

    private fun getCloseableBottom(): Boolean {
        return getCloseDirection() == CloseDirection.DEFAULT || getCloseDirection() == CloseDirection.BOTTOM
    }

    private fun getCloseableTop(): Boolean {
        return getCloseDirection() == CloseDirection.DEFAULT || getCloseDirection() == CloseDirection.TOP
    }
}